command/connect/envoy/pipe-bootstrap/connect_envoy_pipe-bootstrap.go

```diff
@@ -54,22 +59,20 @@ func (c *cmd) Run(args []string) int {
 		return 1
 	}
 
-	_, err = io.Copy(f, os.Stdin)
-	if err != nil {
+	if _, err := buf.WriteTo(f); err != nil {
 		c.UI.Error(err.Error())
 		return 1
 	}
 
-	err = f.Close()
-	if err != nil {
+	if err = f.Close(); err != nil {
 		c.UI.Error(err.Error())
 		return 1
 	}
 
 	// Removed named pipe now we sent it. Even if Envoy has not yet read it, we
 	// know it has opened it and has the file descriptor since our write above
 	// will block until there is a reader.
-	c.UI.Output("Bootstrap sent, unlinking named pipe")
+	c.UI.Warn("Bootstrap sent, unlinking named pipe")
```

Comment:
This has to use `Warn` because both `Output` and `Info` go to stdout, which breaks the tests. Some tests expect to read the stdout of this process to see what was sent to it from the parent process.

This is the 3rd or 4th time where we've noticed that the interface provided by `cli.UI` results in either surprising code, or where it causes bugs because the wrong method is used.

Reply:
I'm actually not exactly sure how those tests work, but this output was definitely breaking the tests, and sending it to stderr fixes them. Sending to stderr seems correct either way, because this is an information log, not expected program output.

Reply:
My guess is this worked before because we weren't passing stdin/stderr through?

I find it a bit confusing that this should be stderr to be honest. It's an expected case not a "warning". I'd argue for just removing it if it's not a useful part of program output or moving it to "debug" or something.

If it's easiest to keep it as a warning, is it worth commenting why it's a warning as I'd assume that was a mistake if I came across this later without this context (which I probably will in a few months 😄 )?

Reply:
> It's an expected case not a "warning"

This is a common misunderstanding of stdout/stderr. Stderr is not simply for warnings, it's for any output that isn't requested by the user. Any output that you wouldn't expect to pipe to another process. All logs at all log levels should go to stderr.

This is why the `UI` interface is a problem, it sends `Info` to stdout.